File: homecontroller_api/src/ticket_service.ts
import { Logger } from './util/logger';
import Exception from './util/exception';
import generateToken from './util/token';

export interface UserTicketData {
    userId: string,
    nodeId: string,
    used: boolean
};

export interface DeviceTicketData {
    userId: string,
    deviceId: string,
    nodeId: string,
    used: boolean
};

export default class TicketService {
    private userTickets: Map<string, UserTicketData>;
    private deviceTickets: Map<string, DeviceTicketData>;

    constructor() {
        this.userTickets = new Map<string, UserTicketData>();
        this.deviceTickets = new Map<string, DeviceTicketData>();
    }

    public createUserTicket(userId: string, nodeId: string): string {
        let ticketData: UserTicketData = {
            userId: userId,
            nodeId: nodeId,
            used: false
        };

        let ticket: string = generateToken(32);
        this.userTickets.set(ticket, ticketData);

        setTimeout((): void => {
            let ticketData = this.userTickets.get(ticket);
            if (ticketData) {
                if (!ticketData.used) {
                    Logger.log('user ticket [' + ticket + '] expired');
                }

                this.userTickets.delete(ticket);
            }
        }, 5000);

        Logger.log('created ticket [' + ticket + '] for user');

        return ticket;
    }


    public checkUserTicket(ticket: string, nodeId: string): UserTicketData {
        if (!this.userTickets.has(ticket)) {
            throw new Exception('bad ticket', 400);
        }

        let ticketData: UserTicketData = this.userTickets.get(ticket)!;

        if (ticketData.used) {
            throw new Exception('ticket used', 400);
        }

        if (ticketData.nodeId !== nodeId) {
            throw new Exception('bad ticket', 400);
        }

        ticketData.used = true;
        this.userTickets.set(ticket, ticketData);

        Logger.log('user [' + ticketData.userId + '] used ticket [' + ticket + ']');

        return ticketData;
    }

    public createDeviceTicket(userId: string, deviceId: string, nodeId: string): string {
        let ticketData: DeviceTicketData = {
            userId: userId,
            deviceId: deviceId,
            nodeId: nodeId,
            used: false
        };

        let ticket: string = generateToken(32);
        this.deviceTickets.set(ticket, ticketData);

        setTimeout((): void => {
            let ticketData = this.deviceTickets.get(ticket);
            if (ticketData) {
                if (!ticketData.used) {
                    Logger.log('device ticket [' + ticket + '] expired');
                }

                this.deviceTickets.delete(ticket);
            }
        }, 5000);

        Logger.log('created ticket [' + ticket + '] for device');

        return ticket;
    }

    public checkDeviceTicket(ticket: string, nodeId: string): DeviceTicketData {        
        if (!this.deviceTickets.has(ticket)) {
            throw new Exception('bad ticket', 400);
        }

        let ticketData: DeviceTicketData = this.deviceTickets.get(ticket)!;

        if (ticketData.used) {
            throw new Exception('ticket used', 400);
        }

        if (ticketData.nodeId !== nodeId) {
            throw new Exception('bad ticket', 400);
        }

        ticketData.used = true;
        this.deviceTickets.set(ticket, ticketData);

        Logger.log('device [' + ticketData.deviceId + '] used ticket [' + ticket + ']');

        return ticketData;
    }
}

File: device_node/src/ticket_service.cpp
#include "ticket_service.h"

#include <random>
#include <iostream>
#include <thread>

#include <homecontroller/util/bits.h>
#include <homecontroller/util/logger.h>
#include <homecontroller/api/info.h>

std::string ticket_service::create_user_ticket(const std::string& user_id) {
    std::lock_guard<std::mutex> lock(m_mutex);

    std::string ticket = hc::util::bits::random_identifier(hc::api::info::TICKET_LENGTH);

    ticket_data data;
    data.m_type = type::USER;
    data.m_user_id = user_id;
    data.m_create_time = std::chrono::system_clock::now();
    data.m_used = false;

    m_tickets.insert(std::make_pair(ticket, data));

    return ticket;
}

std::string ticket_service::create_device_ticket(const std::string& user_id, const std::string& device_id) {
    std::lock_guard<std::mutex> lock(m_mutex);

    std::string ticket = hc::util::bits::random_identifier(hc::api::info::TICKET_LENGTH);

    ticket_data data;
    data.m_type = type::DEVICE;
    data.m_user_id = user_id;
    data.m_device_id = device_id;
    data.m_create_time = std::chrono::system_clock::now();
    data.m_used = false;

    m_tickets.insert(std::make_pair(ticket, data));

    return ticket;
}

bool ticket_service::check_user_ticket(const std::string& ticket, std::string& user_id_ref) {
    std::lock_guard<std::mutex> lock(m_mutex);

    auto mit = m_tickets.find(ticket);
    if (mit == m_tickets.end()) {
        return false;
    }

    if (mit->second.m_type != type::USER || mit->second.m_used) {
        return false;
    }

    user_id_ref = mit->second.m_user_id;

    mit->second.m_used = true;

    return true;
}

bool ticket_service::check_device_ticket(const std::string& ticket, std::string& user_id_ref, std::string& device_id_ref) {
    std::lock_guard<std::mutex> lock(m_mutex);

    auto mit = m_tickets.find(ticket);
    if (mit == m_tickets.end()) {
        return false;
    }

    if (mit->second.m_type != type::DEVICE || mit->second.m_used) {
        return false;
    }

    user_id_ref = mit->second.m_user_id;
    device_id_ref = mit->second.m_device_id;

    mit->second.m_used = true;

    return true;
}

void ticket_service::ticket_expire_loop() {
    m_should_kill_ticket_expire_loop = false;

    while(!m_should_kill_ticket_expire_loop) {
        std::chrono::time_point<std::chrono::system_clock> curr_time = std::chrono::system_clock::now();

        for (auto mit = m_tickets.cbegin(); mit != m_tickets.cend();) {
            if (mit->second.m_used) {
                std::lock_guard<std::mutex> lock(m_mutex);
                m_tickets.erase(mit++);
            }

            else if (std::chrono::duration_cast<std::chrono::milliseconds>(curr_time - mit->second.m_create_time).count() >= 5000) {
                std::lock_guard<std::mutex> lock(m_mutex);
                hc::util::logger::log("ticket [" + mit->first + "] expired");
                m_tickets.erase(mit++);
            } 
            
            else {
                ++mit;
            }
        }

        std::this_thread::sleep_for(std::chrono::milliseconds(500));
    }

}